src/mame/sinclair/atm.cpp

```diff
@@ -13,118 +23,379 @@ Not working because of banking issues.
+		}
+		else
+		{
+			if ((page & DOS7FFD_MASK) && !BIT(page, 1))
+				page = (page & ~1) | SHADOW;
+			page = page & 0x03; // TODO size dependent
+			m_bank_rom[bank]->set_entry(page);
+			views[bank].get().select(0);
+			LOGMEM(" RO(%X>%X)", m_bank_rom[bank]->entry(), page);
+		}
+	}
+	LOGMEM("\n");
+}
 
-	m_screen_location = messram + ((m_port_7ffd_data & 8) ? (7<<14) : (5<<14));
+void atm_state::atm_ula_w(offs_t offset, u8 data)
+{
+	m_br3 = ~offset & 0x08;
+	spectrum_128_state::spectrum_ula_w(offset, data);
+}
 
-	m_bank4->set_base(messram + ((m_port_7ffd_data & 0x07) * 0x4000));
+void atm_state::atm_port_ffff_w(offs_t offset, u8 data)
+{
+	if(!SHADOW)
+		return;
 
-	if (m_beta->started() && m_beta->is_active() && !( m_port_7ffd_data & 0x10 ) )
-		m_rom_selection = 3;
+	if (m_pen2)
+	{
+		m_beta->param_w(data);
+	}
 	else
-		/* ROM switching */
-		m_rom_selection = BIT(m_port_7ffd_data, 4) ;
-
-	/* rom 0 is 128K rom, rom 1 is 48 BASIC */
-	m_bank1->set_base(&m_p_ram[0x10000 + (m_rom_selection<<14)]);
+	{
+		// Must read current ULA value (which is doesn't work now) from the BUS.
+		// Good enough as none-border case is too complicated and possibly none software uses it.
+		u8 pen = get_border_color(m_screen->hpos(), m_screen->vpos());
+		m_palette->set_pen_color(pen,
+			(BIT(~data, 1) * 0xaa) | (BIT(~data, 6) * 0x55),
+			(BIT(~data, 4) * 0xaa) | (BIT(~data, 7) * 0x55),
+			(BIT(~data, 0) * 0xaa) | (BIT(~data, 5) * 0x55));
+	}
 }
 
-void atm_state::atm_port_7ffd_w(uint8_t data)
+void atm_state::atm_port_7ffd_w(offs_t offset, u8 data)
 {
 	/* disable paging */
-	if (m_port_7ffd_data & 0x20)
+	if (BIT(m_port_7ffd_data, 5))
 		return;
 
-	/* store new state */
 	m_port_7ffd_data = data;
+	atm_update_memory();
 
-	/* update memory */
+	m_screen->update_now();
+	m_screen_location = m_ram->pointer() + ((BIT(m_port_7ffd_data, 3) ? 7 : 5) << 14);
+}
+
+void atm_state::atm_port_ff77_w(offs_t offset, u8 data)
+{
+	if (!SHADOW)
+		return;
+
+	m_pen = BIT(offset, 8);
+	m_cpm = BIT(offset, 9);
+	m_pen2 = BIT(offset, 14);
+	LOGMASKED(LOG_VIDEO | LOG_MEM, "PEN %s, CPM %s, PEN2 %s\n", m_pen ? "on" : "off", m_cpm ? "off" : "on", m_pen2 ? "off" : "on");
+	atm_update_memory();
+
+	m_maincpu->set_clock(X1_128_SINCLAIR / 10 * (1 << BIT(data, 3))); // 0 - 3.5MHz, 1 - 7MHz
+
+	int rg = data & 0x07;
+	if ( m_rg ^ rg )
+	{
+		m_rg = rg;
+		atm_update_video_mode();
+	}
+}
+
+void atm_state::atm_port_fff7_w(offs_t offset, u8 data)
+{
+	if (!SHADOW)
+		return;
+
+	u8 bank = offset >> 14;
+	u8 page = (data & 0xc0) | (~data & 0x3f);
+
+	LOGMEM("PEN%s.%s = %X %s%d: %02X\n", (page | DOS7FFD_MASK) ? "+" : "!", BIT(m_port_7ffd_data, 4), data, (page & RAM_MASK) ? "RAM" : "ROM", bank, page & 0x3f);
+	PEN_PAGE(bank) = page;
 	atm_update_memory();
 }
 
-uint8_t atm_state::beta_neutral_r(offs_t offset)
+rectangle atm_state::get_screen_area()
+{
+	switch (m_rg)
+	{
+		case 0b110: // 80x25txt
+			[[fallthrough]];
+		case 0b010: // 640x200
```

Comment:
You don’t actually need the `[[fallthrough]]` hint if there are no statements between the case labels.